### PR TITLE
Add compile time checking for run time locking assertions

### PR DESCRIPTION
**Pull request by practicalswift**
Add compile time checking for all run time locking assertions.

The intention is that literally all `EXCLUSIVE_LOCKS_REQUIRED`/`LOCKS_EXCLUDED`:s added in this PR should follow either directly or indirectly from `AssertLockHeld(…)`/`AssertLockNotHeld(…)`:s already existing in the repo.

Consider the case where function `A(…)` contains `AssertLockHeld(cs_foo)` (without
first locking `cs_foo` in `A`), and that `B(…)` calls `A(…)` (without first locking `cs_main`):
* It _directly_ follows that: `A(…)` should have an `EXCLUSIVE_LOCKS_REQUIRED(cs_foo)` annotation.
* It _indirectly_ follows that: `B(…)` should have an `EXCLUSIVE_LOCKS_REQUIRED(cs_foo)` annotation.

### CODE REVIEW

**Comment by practicalswift** [comment]
WIP: Will move annotations to header files where appropriate.

File: src/txmempool.cpp [hunk 1]
```diff
@@ -502,10 +502,10 @@ void CTxMemPool::removeRecursive(const CTransaction &origTx, MemPoolRemovalReaso
     }
 }
 
-void CTxMemPool::removeForReorg(const CCoinsViewCache *pcoins, unsigned int nMemPoolHeight, int flags)
+void CTxMemPool::removeForReorg(const CCoinsViewCache *pcoins, unsigned int nMemPoolHeight, int flags) EXCLUSIVE_LOCKS_REQUIRED(cs_main, cs, mempool.cs)
 {
     // Remove transactions spending a coinbase which are now immature and no-longer-final transactions
-    LOCK(cs);
+    AssertLockHeld(cs);
```
Comment (MarcoFalke) [inline src/txmempool.cpp]: This 1 line diff should be a separate pull request

**Comment by MarcoFalke** [comment]
Concept ACK

File: src/txmempool.cpp (hunk 1 above)
Comment (practicalswift) [reply src/txmempool.cpp]: That assertion does not seem to hold, so I think it should be removed. I'll investigate.

**Comment by practicalswift** [comment]
@MarcoFalke Updated! Please review :-)

**Comment by laanwj** [comment]
Concept ACK - Unlike the title implies, not all the added run time checking requirements have an associated AssertLockHeld/AssertLockNotHeld (e.g. CoinSelection) is this on purpose?

**Comment by practicalswift** [comment]
@laanwj Thanks for reporting. The annotations for `CoinSelection` and two other functions were incorrect. That is now fixed.

The intention is that literally all `EXCLUSIVE_LOCKS_REQUIRED`:s added in this PR
should follow either directly or indirectly from `AssertLockHeld(…)`/`AssertLockNotHeld(…)`
already existing in the repo. The goal is to transform the run time assertions into compile time checks.

Consider the case where function `A(…)` contains `AssertLockHeld(cs_foo)` (without
first locking `cs_foo`), and that `B(…)` calls `A(…)` (without first locking `cs_main`):
* It _directly_ follows that: `A(…)` should have an `EXCLUSIVE_LOCKS_REQUIRED(cs_foo)` annotation.
* It _indirectly_ follows that: `B(…)` should have an `EXCLUSIVE_LOCKS_REQUIRED(cs_foo)` annotation.

I'll run through all annotations again to verify that all annotations follow directly or indirectly from `AssertLockHeld(…)`/`AssertLockNotHeld(…)`:s.

These are the annotations added – let me know if something looks weird:

```
src/init.cpp
  void ThreadImport(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)

src/keystore.h
  void ImplicitlyLearnRelatedKeyScripts(…) EXCLUSIVE_LOCKS_REQUIRED(cs_KeyStore)

src/net_processing.cpp
  void MaybeSetPeerAsAnnouncingHeaderAndIDs(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  bool TipMayBeStale(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  static bool BlockRequestAllowed(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  void static ProcessGetBlockData(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)
  void static ProcessGetData(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs) LOCKS_EXCLUDED(cs_main)
  bool static ProcessMessage(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)
  static bool SendRejectsAndCheckIfBanned(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  bool PeerLogicValidation::ProcessMessages(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)

src/net_processing.h
  void ConsiderEviction(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)

src/policy/rbf.h
  RBFTransactionState IsRBFOptIn(…) EXCLUSIVE_LOCKS_REQUIRED(pool.cs)

src/qt/transactiondesc.h
  static QString FormatTxStatus(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)

src/qt/transactionrecord.h
  static QList<TransactionRecord> decomposeTransaction(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  void updateStatus(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  bool statusUpdateNeeded() const EXCLUSIVE_LOCKS_REQUIRED(cs_main)

src/rpc/blockchain.cpp
  UniValue blockheaderToJSON(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  void entryToJSON(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)
  UniValue preciousblock(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)
  UniValue invalidateblock(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)
  UniValue reconsiderblock(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)

src/rpc/blockchain.h
  UniValue blockToJSON(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  UniValue blockheaderToJSON(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)

src/rpc/mining.cpp
  UniValue generatetoaddress(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)
  UniValue submitblock(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)

src/rpc/mining.h
  UniValue generateBlocks(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)

src/rpc/rawtransaction.cpp
  UniValue sendrawtransaction(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)

src/test/miner_tests.cpp
  bool TestSequenceLocks(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  BOOST_AUTO_TEST_CASE(CreateNewBlock_validity) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)

src/test/test_bitcoin.cpp
  TestChain100Setup::CreateAndProcessBlock(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)

src/test/txvalidation_tests.cpp
  BOOST_FIXTURE_TEST_CASE(tx_mempool_reject_coinbase, TestChain100Setup) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)

src/test/txvalidationcache_tests.cpp
  ToMemPool(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)
  BOOST_FIXTURE_TEST_CASE(tx_mempool_block_doublespend, TestChain100Setup) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)

src/txmempool.cpp
  void CTxMemPool::removeForReorg(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main, mempool.cs)

src/txmempool.h
  void trackPackageRemoved(…) EXCLUSIVE_LOCKS_REQUIRED(cs)
  std::vector<indexed_transaction_set::const_iterator> GetSortedDepthAndScore() const EXCLUSIVE_LOCKS_REQUIRED(cs)
  void RemoveStaged(…) EXCLUSIVE_LOCKS_REQUIRED(cs)

src/validation.cpp
  bool PreciousBlock(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)
  bool InvalidateBlock(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main, mempool.cs)
  static bool IsCurrentForFeeEstimation() EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  void UpdateMempoolForReorg(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main, mempool.cs)
  static bool CheckInputsFromMempoolAndCache(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  static bool AcceptToMemoryPoolWorker(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main, mempool.cs)
  static bool AcceptToMemoryPoolWithTime(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main, mempool.cs)
  bool AcceptToMemoryPool(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main, mempool.cs)
  static void CheckForkWarningConditions() EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  static void CheckForkWarningConditionsOnNewFork(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  void static InvalidChainFound(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  void CChainState::InvalidBlockFound(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  bool CheckInputs(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  static unsigned int GetBlockScriptFlags(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  bool CChainState::ConnectBlock(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  bool CChainState::ConnectTip(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  bool CChainState::ActivateBestChainStep(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main, mempool.cs)
  CChainState::ActivateBestChain(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)
  bool CChainState::AcceptBlockHeader(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  bool CChainState::AcceptBlock(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  bool LoadChainTip(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)

src/validation.h
  bool ProcessNewBlock(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs) LOCKS_EXCLUDED(cs_main)
  bool LoadExternalBlockFile(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)
  bool ActivateBestChain(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs) LOCKS_EXCLUDED(cs_main)
  bool AcceptToMemoryPool(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main, mempool.cs)
  bool CheckFinalTx(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  bool TestLockPointValidity(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  bool CheckSequenceLocks(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main, mempool.cs)
  bool TestBlockValidity(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  bool PreciousBlock(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)
  bool InvalidateBlock(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main, mempool.cs)
  bool ResetBlockFailureFlags(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  bool LoadMempool() EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)

src/validationinterface.h
  void SyncWithValidationInterfaceQueue() LOCKS_EXCLUDED(cs_main)

src/wallet/feebumper.cpp
  static feebumper::Result PreconditionChecks(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main, wallet->cs_wallet)

src/wallet/rpcdump.cpp
  void ImportScript(…) EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet)
  void ImportAddress(…) EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet)
  UniValue ProcessImport(…) EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet)

src/wallet/rpcwallet.cpp
  void WalletTxToJSON(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  UniValue ListReceived(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  void ListTransactions(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  UniValue generate(…) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)

src/wallet/test/wallet_tests.cpp
  BOOST_AUTO_TEST_CASE(coin_selection_tests) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  BOOST_AUTO_TEST_CASE(ApproximateBestSubset) EXCLUSIVE_LOCKS_REQUIRED(cs_main)

src/wallet/wallet.cpp
  bool CWalletTx::AcceptToMemoryPool(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main, mempool.cs)

src/wallet/wallet.h
  int GetDepthInMainChain(…) const EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  int GetDepthInMainChain() const EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  bool IsInMainChain() const EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  int GetBlocksToMaturity() const EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  CAmount GetCredit(…) const EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  CAmount GetImmatureCredit(…) const EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  CAmount GetAvailableCredit(…) const EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  CAmount GetImmatureWatchOnlyCredit(…) const EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  CAmount GetAvailableWatchOnlyCredit(…) const EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  bool IsTrusted() const EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  bool RelayWalletTransaction(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  bool AcceptToMemoryPool(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  bool SelectCoins(…) const EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  void SyncTransaction(…) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  void DeriveNewChildKey(…) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  bool AddWatchOnly(…) override EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  void LoadKeyPool(…) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  bool CanSupportFeature(…) const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  void AvailableCoins(…) const EXCLUSIVE_LOCKS_REQUIRED(cs_main, cs_wallet)
  bool IsSpent(…) const EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  bool IsLockedCoin(…) const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  void LockCoin(…) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  void UnlockCoin(…) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  void UnlockAllCoins() EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  void ListLockedCoins(…) const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  CPubKey GenerateNewKey(…) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  bool AddKeyPubKey(…) override EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  bool AddKeyPubKeyWithDB(…) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  bool LoadKeyMetadata(…) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  bool LoadScriptMetadata(…) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  bool LoadMinVersion(…) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  void UpdateTimeFirstKey(…) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  bool AddWatchOnly(…) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  bool RemoveWatchOnly(…) override EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  void GetKeyBirthTimes(…) const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  int64_t IncOrderPosNext(…) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  bool AccountMove(…) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  bool AddToWalletIfInvolvingMe(…) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  void ResendWalletTransactions(…) override EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  std::vector<uint256> ResendWalletTransactionsBefore(…) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  bool SignTransaction(…) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  size_t KeypoolCountExternalKeys() EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  void MarkReserveKeysAsUsed(…) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  std::set<std::set<CTxDestination>> GetAddressGroupings() EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  std::map<CTxDestination, CAmount> GetAddressBalances() EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  DBErrors ZapSelectTx(…) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  unsigned int GetKeyPoolSize() EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  bool HasWalletSpend(…) const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
  void BlockUntilSyncedToCurrentChain() LOCKS_EXCLUDED(cs_main, cs_wallet)

src/wallet/walletdb.cpp
  bool ReadKeyValue(…) EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet)
```

**Comment by Sjors** [comment]
Sadly that still didn't catch my [mistake here](https://github.com/bitcoin/bitcoin/pull/12565/files#r171235800), but I assume this is a step in that direction.

Lightly tested on macOS (no change, as expected).

@practicalswift maybe you can put that overview and additional explanation ("directly follows that...") in the commit message or PR description?

I didn't see a `EXCLUSIVE_LOCKS_REQUIRED ` in `LoadChainTip`; I assume that's done in a some dependent function?

The `EXLCUSIVE_LOCKS...` stuff in `validation.h` is not repeated in `validation.cpp`, that's intentional?

Should violations merely throw warnings?

**Comment by MeshCollider** [comment]
Concept ACK

**Comment by practicalswift** [comment]
@Sjors No intention to repeat annotations :-)  Which specific annotations did you see repeated in `validation.cpp`? Did you check them in the files or just in the summary?

Violations are checked by Travis and cause a Travis build failure if found – see [`configure.ac`](https://github.com/bitcoin/bitcoin/blob/f13d756cdd9445d1478a70c5387df2a444045f6e/configure.ac#L251-L257).

**Comment by practicalswift** [comment]
@Sjors Did you mean that you didn't see any `AssertLockHeld(…)` in `LoadChainTip`? :-)

**Comment by Sjors** [comment]
> Which specific annotations did you see repeated in validation.cpp

None, so that's all good. I just found it slightly confusing that some annotations are in .h files and others in .cpp files.

> Did you mean that you didn't see any `AssertLockHeld(…)` in `LoadChainTip`?

Yes, sorry. To more specific, no `AssertLockHeld(mempool.cs)`. But if you're happy and the compiler is happy, I'm happy :-)

**Comment by practicalswift** [comment]
@Sjors The annotations should be in the header files, but not all functions are exported so that's why some annotations end up in the `.h` files.

`LoadChainTip` is annotated with `EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)` since it (indirectly) calls `CheckSequenceLocks` (which contains `AssertLockHeld(mempool.cs)`) without first locking `mempool.cs`.

The call stack looks like this:

```
LoadChainTip
  ActivateBestChain
    ActivateBestChainStep
      UpdateMempoolForReorg
        AcceptToMemoryPool
          AcceptToMemoryPoolWithTime
            AcceptToMemoryPoolWorker
              CheckSequenceLocks
```

**Comment by sipa** [comment]
Concept ACK

**Comment by practicalswift** [comment]
Rebased!

**Comment by practicalswift** [comment]
Rebased again!

**Comment by practicalswift** [comment]
@laanwj This PR is a subset of #11226 (closed in favour of smaller locking PR:s) which had a 0.17.0 milestone (originally a 0.16.0 milestone :-)). Would it be possible to get a 0.17.0 for this PR?

Of the locking PR:s this should be the best candidate for early merge. It contains only locking assertions we are already making but using `AssertLockHeld(…)` (directly or indirectly). Since this PR doesn't change runtime behaviour (compile-time checks only) it should be low risk.

Having this PR merged would make the other locking assertions PR:s much smaller since this PR contains the bulk of the assertions.

Keeping this PR rebased is getting a bit time consuming, so let me know if I can do anything to facilitate merging :-)

**Comment by practicalswift** [comment]
Rebase number 3 performed :-)

**Comment by Sjors** [comment]
Can confirm 9d3e5374e30cd2b1c04c0bb45da24236eb3b2b37 still compiles on macOS.

@TheBlueMatt and @promag recently worked on concurrency related bugs in `init.cpp`, so they might be able to review.

**Comment by practicalswift** [comment]
Rebase number 4 performed :-)  Please review :-)

**Comment by practicalswift** [comment]
@laanwj This PR is getting a bit heavy to be kept up-to-date. Do you think this PR has a chance of getting a 0.17.0 milestone (like #11226 had before getting closed in favour of this and other smaller PR:s)? :-)

File: src/rpc/rawtransaction.cpp [hunk 2]
```diff
@@ -1133,7 +1133,7 @@ UniValue sendrawtransaction(const JSONRPCRequest& request)
     return hashTx.GetHex();
 }
 
-UniValue testmempoolaccept(const JSONRPCRequest& request)
+UniValue testmempoolaccept(const JSONRPCRequest& request) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)
```
Comment (MarcoFalke) [inline src/rpc/rawtransaction.cpp]: I fail to see how this could compile, since we never `LOCK(mempool.cs)` before calling into this, no?

File: src/test/miner_tests.cpp [hunk 3]
```diff
@@ -202,7 +202,7 @@ void TestPackageSelection(const CChainParams& chainparams, CScript scriptPubKey,
 }
 
 // NOTE: These tests rely on CreateNewBlock doing its own self-validation!
-BOOST_AUTO_TEST_CASE(CreateNewBlock_validity)
+BOOST_AUTO_TEST_CASE(CreateNewBlock_validity) EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)
```
Comment (MarcoFalke) [inline src/test/miner_tests.cpp]: Same here

File: src/txmempool.cpp [hunk 4]
```diff
@@ -502,7 +502,7 @@ void CTxMemPool::removeRecursive(const CTransaction &origTx, MemPoolRemovalReaso
     }
 }
 
-void CTxMemPool::removeForReorg(const CCoinsViewCache *pcoins, unsigned int nMemPoolHeight, int flags)
+void CTxMemPool::removeForReorg(const CCoinsViewCache *pcoins, unsigned int nMemPoolHeight, int flags) EXCLUSIVE_LOCKS_REQUIRED(cs_main, mempool.cs)
 {
     // Remove transactions spending a coinbase which are now immature and no-longer-final transactions
     LOCK(cs);
```
Comment (MarcoFalke) [inline src/txmempool.cpp]: It seem the lock is taken here, but you require `EXCLUSIVE_LOCKS_REQUIRED(mempool.cs)`

File: src/validation.cpp [hunk 5]
```diff
@@ -468,7 +467,7 @@ static bool IsCurrentForFeeEstimation()
  * and instead just erase from the mempool as needed.
  */
 
-void UpdateMempoolForReorg(DisconnectedBlockTransactions &disconnectpool, bool fAddToMempool)
+void UpdateMempoolForReorg(DisconnectedBlockTransactions &disconnectpool, bool fAddToMempool) EXCLUSIVE_LOCKS_REQUIRED(cs_main, mempool.cs)
```
Comment (MarcoFalke) [inline src/validation.cpp]: What about adding `static` to functions that have their lock annotations added in the cpp file and not in the header?

**Comment by practicalswift** [comment]
@MarcoFalke Updated. I've now tried to address your feedback. Turns out that I was a bit too strict with the locking requirements for `mempool.cs`, so the updated diff is significantly smaller and hopefully easier to review. Please re-review :-)

**Comment by MarcoFalke** [comment]
utACK (could use `::mempool.cs` instead of `mempool.cs`, since it is referring to the global, I guess)

Also, you could split up the `cs_wallet` and `::mempool.cs` into two separate prs. I have the feeling it is easier to get stuff in that only touches a single "module".

**Comment by practicalswift** [comment]
@MarcoFalke Thanks for the utACK!


Added a commit which changed from `mempool.cs` to `::mempool.cs`.

I've now split up this PR into four separate PRs – one per lock:
* `cs_KeyStore`: #13077
* `::mempool.cs`: #13080
* `cs_wallet`: #13081
* `cs_main`: #13083

Please review these four :-)
